[Week1] ToDo REST API 만들기

Miniminis opened the pull request:
# Week1 - ToDo REST API 만들기

## 과제 목표
Java를 이용해서 ToDo REST API를 만든다.

![API](https://user-images.githubusercontent.com/14071105/103476206-0456f280-4df7-11eb-89c4-d61845ef45ec.png)

## 공식적인 과제통과 조건

- [x] 모든 API 테스트를 통과
- [x] 주간회고 공유

## 개인적인 과제 통과 조건

- [ ] 1주차 배운 내용들에 대해서 블로그에 '나의 언어로' 글쓰고 - 공유
   - [ ] java 개발환경 구축하기 : jdk, gradle, intelliJ
   - [ ] REST API : HTTP, response status code, request method, resource, httpie
   - [ ] Jackson : JSON, Jackson
- [x] 조금이라도 헷갈리는 개념들 질문 - 정리 - 공유


## 실행하기

```bash
./gradlew run
```

## 테스트

### 설치하기

```bash
$ cd tests
$ npm install
```

### 테스트 실행하기

테스트는 실제로 동작하는 서버에 테스트하므로 서버가 동작하고 있는 상태여야 올바르게 동작합니다.

```bash
$ npm run test
```


## 작업 목록

### 프로젝트 환경설정

- [ ] 자바 15버전을 설치한다

### 웹서버 세팅

- [x] Http server 만들기 : no connection -> 404 -> 500 넘어가는 흐름 파악
- [x] handler 생성
- [x] header, body response & log 세팅
- [x] models package 생성 > Task model 생성
- [x] Jackson lib dependency
- [x] task to json / json to task 공통 매소드 만들기

### ToDo 목록 얻기 - `GET /tasks`

- [x] tasks 목록 응답

### ToDo 상세 조회하기 - `GET /tasks/{id}`

- [x] path 의 id에 해당하는 task 내용 응답
- [x] id가 0 이하일 때, 오류메시지 응답

### ToDo 생성하기 - `POST /tasks`

- [x] 마지막 id + 1에 해당하는 신규 task 추가
- [x] task title 형식이 안맞을 때 오류 응답 : 문자가 아니거나 비어있으면 안됨

### ToDo 제목 수정하기 - `PUT/PATCH /tasks/{id}`

- [x] id에 맞는 task 찾아서 제목 수정하기
- [x] 예외처리 - id 에 맞는 task 가 존재하지 않을 때 오류 응답
- [x] 예외처리 - 수정하려는 값이 적절하지 않을 때 오류 응답 : 문자가 아니거나 비어있으면 안됨

### ToDo 삭제하기 - `DELETE /tasks/{id}`

- [x] id에 맞는 값을 찾아서 삭제한다
- [x] 예외처리 - id 에 맞는 task 가 존재하지 않을 때 오류 응답

### tests > 전체 테스트

- [x] 전체 테스트를 통과한다


## 추가 작업/공부 목록 (from feedback)

### 2022.03.22

- [x] git PR 프로세스 제대로 알고 있는지 한번 더 확인하기

### 2022.03.22

- [ ] Object Mapper 이용하지 않고 json - task 간 parsing 하도록 수정하기
- [x] 경로 Path validation - GET /tasks/ 처럼 뒤쪽에 `/` 붙은 경로들 대응하기
- [x] 일반 숫자, 문자들 상수로 분리하기
- [x] `import *` 처럼 전체 import 사용하지 않고 필요한 라이브러리만 import 하기

### 2022.03.23

- [x] git PR 상단에 TODO list 만드는 법 조사 후 적용해보기
- [x] HttpMethod - [http method 정의된 RFC 문서](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3) 링크 주석으로 남겨둬서 편하게 읽을 수 있도록 하기
- [x] HttpMethod - [응답 메시지 RFC 7231에 명시된](https://datatracker.ietf.org/doc/html/rfc7231#section-6.1) Reason-Phrase를 적어주는 방식으로 수정하기
- [x]  HttpMethod - response code : http method 에 정의해준 enum 활용하기
- [ ] TodoHttpHandler > matchPathDepthOne, matchPathDepthTwo
   - [x] `PATH_REGEX` : 매소드 인자로 받게 고치기
   - [x] 매소드 이름 좀 더 일반적인 의미로 수정하기
   - [x] 해당 매소드 다른 클래스로 옮기고 public 으로 접근제한 풀기
   - [x] Junit test code 작성하기
   - [ ] Junit > @Nested 사용해서 계층형 테스트 코드 작성하기

### 2022.03.25

- [x] REST API - POST/PUT 매소드에 대해서 response 관례/규약 질문 후 내용 정리하기

### 2022.03.26

- [x] [JavaDoc 태그 사용해보기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835233263)
   - [x] [필요없는 문서 주석 삭제하기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835233823)
- [ ] [reqBody에 재할당을 하지 않고 문제를 해결할 수 있는 방법을 고민하기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835235129)
- [ ] TodoHttpHandler.java > [Path 처리 > 2 : 코드만 보고 논리적으로 근거를 떠올릴 수 있도록 개선하기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835236008)
- [ ] Task.java > [Null or Empty 검사하는 매소드에 JavaDoc 태그 사용해보기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835237287)
- [x] TaskList.java > ? : 를 사용하지 말고 평범한 if문을 사용하는 방식으로 바꿔보기 
- [ ] [Logger 관련 RFC 5242 문서 읽기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835239690)
   - [x] JavaDoc 태그로 주석달기
- [x] [TodoHttpHandlerTest.java > 파일 마지막 라인에 공백 추가하기 & 종립님의 관련 글 보고 공부하기](https://github.com/CodeSoom/spring-week1-assignment-1/pull/82#discussion_r835240867)

Miniminis committed:
docs 요구사항 분석 후 작업목록 및 순서 작성

Miniminis committed:
feat HttpServer 생성 완료

- http server 생성
- context 및 handler 생성
- request, response log 세팅
- header response, body content response 완료

Miniminis committed:
feat Task model 생성 및 task-json parsing 공통매서드 생성

- 할일 관리를 위한 Task 클래스 생성
- json 타입으로 request, response를 편해게 하기 위해 task와 json 간 parsing 공통 매서드 생성

Miniminis commented:
Git PR 관련 질문이 있습니다! 

올려주신 Github 관련 내용을 읽고 나서 과제를 진행했는데, 제가 이해한 내용은 아래와 같았습니다. 
1. 코드숨 repository 에서 내 계정으로 fork 한다. (CodeSoom/spring-week1-assignment-1/main -> Miniminis/spring-week1-assignment-1/main)
3. 내 계정으로 fork 된 repository 를 로컬로 clone 한다. (Miniminis/spring-week1-assignment-1/main -> Local/spring-week1-assignment-1/main)
4. clone 한 프로젝트에서 작업을 위한 새로운 브랜치(todo-rest-api)를 생성한다. (Local/spring-week1-assignment-1/main -> Local/spring-week1-assignment-1/todo-rest-api)
5. 브랜치에서 작업 후 내 계정의 원격 repository 로 push 한다. (Local/spring-week1-assignment-1/todo-rest-api -> Miniminis/spring-week1-assignment-1/todo-rest-api) 
6. 내 원격 repository의 브랜치(todo-rest-api)에서 코드숨 원격 repo의 내 계정 브랜치(miniminis)에 PR 요청을 한다. (Miniminis/spring-week1-assignment-1/todo-rest-api → CodeSoom/spring-week1-assignment-1/miniminis)

그런데 다른 분들의 PR 내역들을 보면 [본인계정]:[본인계정브랜치]-> 코드숨:[본인계정 브랜치] 와 같이 되어있어서 혹시 제가 잘못 이해하고 있나해서 확인차 질문드립니다 :)

johngrib commented:
> 그런데 다른 분들의 PR 내역들을 보면 [본인계정]:[본인계정브랜치]-> 코드숨:[본인계정 브랜치] 와 같이 되어있어서 혹시 제가 잘못 이해하고 있나해서 확인차 질문드립니다 :)

맞게 잘 하셨어요.

johngrib commented on README.md:
````diff
@@ -38,3 +38,39 @@ $ npm install
 $ npm run test
 ```
+
+## 작업목록 및 순서
+1. 초기 세팅
+   1. Http server 만들기 : no connection -> 404 -> 500 넘어가는 흐름 파악
+   2. handler 생성
+   3. header, body response & log 세팅
+   4. models package 생성 > Task model 생성
+      1. id
+      2. title 
+   5. Jackson lib dependency 
+   6. task to json / json to task 공통 매소드 만들기
+2. ToDo 목록 얻기 & 테스트
+   1. 경로 : GET /tasks
+   2. 기능 : tasks 목록 응답
+3. ToDo 상세 조회하기 & 테스트
+   1. 경로 : GET /tasks/{id}
+   2. 기능 : path 의 id에 해당하는 task 내용 응답 
+   3. 예외처리
+      1. id 가 0 이하일때 오류메시지 응답
+4. ToDo 생성하기 & 테스트
+   1. 경로 : POST /tasks
+   2. 기능 : 마지막 id + 1에 해당하는 신규 task 추가
+   3. 예외처리
+      1. task title 형식이 안맞을 때 오류 응답 : 문자가 아니거나 비어있으면 안됨
+5. ToDo 제목 수정하기 & 테스트
+   1. 경로 : PUT/PATCH /tasks/{id}
+   2. 기능 : id에 맞는 task 찾아서 제목 수정하기
+   3. 예외처리
+      1. id 에 맞는 task 가 존재하지 않을 때 오류 응답
+      2. 수정하려는 값이 적절하지 않을 때 오류 응답 : 문자가 아니거나 비어있으면 안됨ㄴ
+6. ToDo 삭제하기 & 테스트
+   1. 경로 : DELETE /tasks/{id}
+   2. 기능 : id에 맞는 값을 찾아서 삭제한다. 
+   3. 예외처리
+      1. id 에 맞는 task 가 존재하지 않을 때 오류 응답
+7. tests > 전체 테스트
````
오 계획을 세워놓고 작업하시는군요. 이 계획은 작업하면서 하나씩 지워나갈 생각이신가요?

Miniminis replied:
원래 작업 시작전에 처음부터 끝까지 한번 그림을 그려보는 시간을 갖곤 합니다! 지워나가면 좋긴 한데, 사실 고민이 되는 부분입니다! MD file 의 numbering format은 intelliJ 에서 편하게 지원해줘서 막 썼다가 지우기가 편리한데, check 스타일은 넘버링만큼 편하게 썼다 지우기가 좀 힘들어서요! 

이 부분은 사실 추후에 변동 가능성이 낮긴 해서 초반에 numbering 으로 편하게 썼다가 지워서 어느정도 초안을 작성한 뒤, check 스타일로 포멧을 바꾸는 건 어떨지 생각하고 있습니다 ㅎㅎ

johngrib replied:
github PR 본문(PR 제일 위에 줄줄이 설명을 적을 수 있는 곳이요)에서도 TODO 체크리스트를 만들 수 있어요. 한번 조사해 보시고 그 방법을 써보시는 것도 좋을 것 같습니다 :-)

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,10 +1,25 @@
 
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
+        InetSocketAddress address = new InetSocketAddress(8000);
```
`8000`을 상수로 분리하는 방법도 고려해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,69 @@
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+
+import java.io.*;
```
`import *` 방식으로 `import`하지 않는 것이 좋은 습관입니다. 어떤 것을 `import`하는지 명확히 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,69 @@
+
+public class TodoHttpHandler implements HttpHandler {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
`ObjectMapper` 없이 해보면 어떨까요?

johngrib reviewed (commented):
반갑습니다!

Miniminis committed:
feat todo 목록 조회 구현 완료

- http methods enum 클래스로 관리

Miniminis committed:
refactor json - task parsing 매서드 별도 클래스로 분리

- JsonParser 클래스를 통해 handler에 섞여있던 parsing 로직을 분리함

* todo
- Object Mapper 이용하지 않고 json - task 간 parsing 하도록 수정하기

Miniminis committed:
refactor 요청 및 응답내용 로깅을 위한 Logger class 생성

- httpHandler 에서 로깅 관련 로직 Logger 클래스로 분리

Miniminis committed:
refactor 전체 구조 MVC 로 리펙토링

- 요청을 받고 요청데이터 validation 을 처리하는 controller, 비즈니스 로직을 처리하는 service, 데이터를 관리하는 model 에 코드 분리
- TaskList 컬랙션 객체 생성하여 데이터 관련 처리 클래스 내부에서 할 수 있도록 함

Miniminis committed:
feat Todo 생성하기 API 초안 완료

- path 에 맞게 호출하면 새로운 task 생성되어 tasks 배열에 추가되도록 로직구성 완료
- TODO)
	- 필요한 예외처리
	- 데이터의 validation 추가

Miniminis committed:
refactor request 경로 validation 추가한 공통로직 생성

- 경로 뒤에 "/" 붙는 등 equals 매소드만으로 대응하기 어려웠던 부분 개선

Miniminis committed:
feat ToDo 상세 조회하기 초안 작성 완료

Miniminis committed:
feat Todo 수정 및 삭제

- 최소한의 validation 만 유지하여 수정, 삭제 API 생성

Miniminis committed:
refactor 숫자, 문자열 모두 상수 처리

Miniminis committed:
refactor `import *` 처럼 전체 import 사용하지 않고 필요한 라이브러리만 import 하도록 수정

Miniminis commented:
오늘 전체 프로젝트 구조를 다시 정리하고 전체 API의 초안 버전을 완성했습니다. 
정말 돌아가는 것만 확인한 수준이라서 아직 개선해야할 부분이 많을 것 같습니다! 
내일 데이터 관련된 validation 을 보완하고, 남겨주신 피드백 기준으로 다시 리펙토링 진행하겠습니다 🙌

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpMethod.java:
```diff
@@ -0,0 +1,19 @@
+package com.codesoom.assignment.enums;
+
+public enum HttpMethod {
```
주석으로 HTTP 메소드가 정의된 RFC 문서 링크를 남겨두는 것도 괜찮겠죠. HTTP는 유명한 프로토콜이라 이런 링크를 생략하는 것이 당연한 일일 수 있지만, 만약 새로운 규약이 있는 기술을 사용하게 된다면 근원이 되는 공식 문서, RFC 문서를 링크해두면 좋습니다. 꼭 습관으로 만들어 두세요. 링크하는 김에 읽어보는 것도 좋을 거고요.

https://datatracker.ietf.org/doc/html/rfc7231#section-4.3

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpStatusCode.java:
```diff
@@ -0,0 +1,29 @@
+
+public enum HttpStatusCode {
+
+    OK(200, "정상적으로 처리되었습니다."),
+    CREATED(201, "정상적으로 생성되었습니다."),
+
+    BAD_REQUEST(400, "잘못된 요청입니다."),
+    NOT_FOUND(404, "결과를 찾을 수 없습니다."),
+
```
한국어를 사용하는 것도 경우에 따라 좋은 아이디어일 수 있어요. 그러나 여기에서는 RFC 7231에 명시된 Reason-Phrase를 적어주는 방식으로 일단 구현해 보도록 합시다. 백엔드 구현은 외부로 내보내야 하는 문자열에 한국어를 사용하는 데에 대해서는 신중할 필요가 있습니다. 이 API 클라이언트가 한국인이 아닐 수도 있거든요. (사람이 아닐 수도 있습니다).

https://datatracker.ietf.org/doc/html/rfc7231#section-6.1

![image](https://user-images.githubusercontent.com/1855714/159697763-26e26d4b-f359-40c1-a4b5-43257aed461a.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,104 @@
+
+
+        httpExchange.getResponseHeaders().set("Content-Type", "application/json");
+        httpExchange.sendResponseHeaders(200, resBody.getBytes(StandardCharsets.UTF_8).length);
```
`200` 대신 정의해둔 enum을 사용하기 좋아 보입니다 :-)

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,104 @@
+        return Long.valueOf(pathElements[index]);
+    }
+
+    private boolean matchPathDepthOne(String reqPath, String mappedPath) {
+        String[] pathElements = reqPath.split(PATH_REGEX);
+        String[] mappedPathElements = mappedPath.split(PATH_REGEX);
+
+        return pathElements.length == mappedPathElements.length
+                && pathElements[1].equals(mappedPathElements[1]);
+    }
```
`PATH_REGEX`를 필드 멤버를 참조하는 방식으로 사용하셨군요. 이렇게 해보면 어떨까요?

- `PATH_REGEX`를 메소드 인자로 받게 고칩니다.
- 메소드 본문을 읽고 열심히 생각해서 `matchPathDepthOne`의 이름을 좀 더 일반적인 의미로(다양한 용도에 사용할 수 있는 이름으로) 수정합니다.
- 메소드를 다른 클래스로 옮기고 `public`으로 접근 제한을 풉니다.
- 시간이 충분하다면 이 메소드의 테스트 코드를 Junit으로 작성합니다.
- 시간이 그래도 충분하다면 Junit의 `@Nested`를 사용해 계층형 테스트 코드를 작성합니다.

Miniminis committed:
refactor httpMethod 클래스에 RFC 문서 링크 주석 추가

- 정의된 내용의 근거가 되는 공식문서(RFC문서)링크를 추가하는 습관 들이기

Miniminis committed:
refactor httpCode 에 맞춰서 response 변경

- 모든 api의 http response code, body 를  HttpStatusCode enum 내용 적용하여 변경
- httpStatusCode enum > RFC 7231에 명시된 Reason-Phrase를 적어주는 방식으로 수정
- POST, PUT, DELETE 시, response 로 Task 타입 반환하도록 변경
- Task > 기본 setter 쓰지 않고 행동의도가 드러나는 이름으로 변경
	- setId > nextId
	- setTitle > editTaskTitle

Miniminis commented:
## 2022.03.25 
### 작업내용
아침까지 종립님이 남겨주신 http method 관련 코멘트들을 검토한 뒤 적용하였습니다. 상세히 코멘트 남겨주셔서 감사합니다 :) 

### 남은작업
- Object Mapper 이용하지 않고 json - task 간 parsing 하도록 수정하기
- TodoHttpHandler > matchPathDepthOne, matchPathDepthTwo 매소드 이름 좀 더 명확하게 고민해보기, 클래스 분리하기
- git PR 상단에 TODO list 만드는 법 조사 후 적용해보기

Miniminis commented:
## 질문

POST, PUT의 경우, 추가나 수정된 내용을 response 해주는게 관례인가요? 현업에서 할때에는 201, 200 정도로 response 코드나 메시지만 내보내주고, 실제 확인은 다시 GET 매소드를 통해서 하고 있었는데, 테스트에 계속 실패하던 부분을 찬찬히 살펴보던 중, POST 와 PUT 매소드에서 response 로 Task 객체를 내보내주지 않아서였다는 것을 발견해서 여쭤봅니다! 

p.s. 어제도 깃 장애가 있었던 것으로 알고 있습니다! 어려운 상황에서도 피드백 주셔서 감사합니다 👍🏾

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpMethod.java:
```diff
@@ -0,0 +1,23 @@
+package com.codesoom.assignment.enums;
+
+/**
+ * https://datatracker.ietf.org/doc/html/rfc7231#section-4.3
+ * */
```
기왕 하는 김에 JavaDoc 태그를 사용해 봅시다. 이렇게 해 보세요. 그리고 IntelliJ 에서 HttpMethod 클래스 이름 위에 마우스를 올려보고 어떻게보이는지도 확인해 보세요.
```suggestion
/**
 * @see <a href="https://datatracker.ietf.org/doc/html/rfc7231#section-4.3">RFC 7231 section 4.3</a> 
 */
```

Miniminis replied:
<img width="495" alt="image" src="https://user-images.githubusercontent.com/44446425/160235040-d5154b06-ae5b-4673-9b4a-d7330325f84d.png">

대박! 엄청신기하네요 👍🏾 이런게 있는 줄 몰랐습니다...!

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpStatusCode.java:
```diff
@@ -0,0 +1,32 @@
+package com.codesoom.assignment.enums;
+
+/**
+ * https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
+ * https://datatracker.ietf.org/doc/html/rfc7231#section-6.1
+ * */
```
MDN 문서도 결국은 RFC를 읽고 쓴 문서죠. 

```suggestion
/**
 * https://datatracker.ietf.org/doc/html/rfc7231#section-6.1
 * */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,119 @@
+                .lines()
+                .collect(Collectors.joining("\n"));
+
+        Task reqBody = null;
```
잘 하셨습니다. 이제 `reqBody`에 재할당을 하지 않고 문제를 해결할 수 있는 방법을 고민해 보세요. 가능한 한 재할당을 하지 않는 것이 좋습니다. (물론 재할당을 안 하는 것보다는 재할당을 하면서 먼저 구현할 수 있는 것이 더 좋습니다. 먼저 재할당으로 만들고, 그 다음에 재할당을 하지 않게 고치는 연습을 하는 방법도 좋아요.)

johngrib commented on app/src/main/java/com/codesoom/assignment/handlers/TodoHttpHandler.java:
```diff
@@ -0,0 +1,119 @@
+        String[] pathElements = reqPath.split(PATH_REGEX);
+
+        return matchPathDepthOne(reqPath, mappedPath)
+                && pathElements[2].chars().allMatch(Character::isDigit);
```
`2`가 뜬금없어 보입니다. 좀 더 자세히 말하자면 이 메소드 내의 코드만 읽고는 `2`의 논리적인 근거를 떠올리기가 어려워요. 상상으로만 `2`가 어떤 이유로 여기에 나타나게 되었는지를 알 수 있죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,54 @@
+        return Objects.equals(id, taskId);
+    }
+
+    public boolean hasValidContent() {
+        return title != null && !title.isEmpty();
+    }
```
이 메소드에 JavaDoc 주석을 달아보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/models/TaskList.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public Task addTask(Task newTask) {
+        long lastIdx = tasks.size() == 0 ? 0L : tasks.get(tasks.size()-1).getId();
```
`? :` 를 사용하지 말고 평범한 `if`문을 사용하는 방식으로 바꿔 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/Logger.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.net.URI;
+
+public class Logger {
+
+    public void logRequest(String requestMethod,
+                           URI requestURI,
+                           String requestPath,
+                           Task requestBody) {
+
+        System.out.println("=====> REQUEST");
+        System.out.println("requestMethod = " + requestMethod);
+        System.out.println("requestURI = " + requestURI);
+        System.out.println("requestPath = " + requestPath);
+
+        if (requestBody != null) {
+            System.out.println("requestBody = " + requestBody);
+        }
+    }
+
+    public void logResponse(BaseResponse response) {
+        System.out.println("=====> RESPONSE");
+        System.out.println("response = " + response.toString());
+        System.out.println();
+    }
+
+}
```
`Logger`를 정의하셨군요! https://datatracker.ietf.org/doc/html/rfc5424 RFC 5424 를 읽을 때가 됐습니다. (훑어보기만 하셔도 괜찮아요)

johngrib commented on app/src/test/java/com/codesoom/assignment/handlers/TodoHttpHandlerTest.java:
```diff
@@ -0,0 +1,20 @@
+        System.out.println("Arrays.toString(strings) = " + Arrays.toString(strings));
+    }
+
+}
```
파일 마지막에 공백 라인 하나를 추가해 주세요. 파일이 `}` 와 같이 공백이 아닌 문자로 끝나게 되면 github에서 다음과 같이 경고가 나옵니다.

![image](https://user-images.githubusercontent.com/1855714/160124050-a012f56d-0aac-458b-b22f-aca1f9a07360.png)


이것이 문제가 되는 이유는 POSIX 표준에서 Line을 "개행 문자로 끝나는 0개 이상의 개행 문자가 아닌 문자들의 시퀀스"으로 정의하고 있기 때문입니다.

만약 파일을 작성할 때마다 마지막에 공백 라인을 추가하는 것이 귀찮다면, IntelliJ에 관련 설정을 해주는 방법이 있습니다.

다음은 제 블로그에 설정 방법을 기록한 것입니다. 참고해주세요.
https://johngrib.github.io/wiki/intellij/#%ED%8C%8C%EC%9D%BC-%EB%A7%88%EC%A7%80%EB%A7%89%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-1%EA%B0%9C%EC%9D%98-%EA%B0%9C%ED%96%89-%EB%AC%B8%EC%9E%90-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0

Miniminis committed:
feat README 문서에 세부 진척사항 확인할 수 있도록 형식 수정

Miniminis committed:
feat @see 태그 사용하여 클래스/매소드와 관련있는 문서 주석에 명시하기

Miniminis committed:
feat TaskList.java > ? : 를 사용하지 말고 평범한 if문을 사용하는 방식으로 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/models/TaskList.java:
```diff
@@ -0,0 +1,65 @@
+        return getTask(newTask.getTitle());
+    }
+
+    public Task getTask(Long taskId) {
+        for (Task task : tasks) {
+            if (task.equalTaskId(taskId)) {
+                return task;
+            }
+        }
+
+        return null;
+    }
```
Task를 저장할 때 List가 적절한지에 대해 고민해볼 필요가 있습니다. 다른 자료구조로 바꾼다면 무엇이 좋을까요? 그리고 그 이유는 무엇일까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TodoService.java:
```diff
@@ -0,0 +1,51 @@
+            return new BaseResponse<>(HttpStatusCode.NOT_FOUND);
+        }
+
+        return new BaseResponse<>(HttpStatusCode.OK, selected.editTaskTitle(task.getTitle()));
```
`task`가 `null`인 경우도 방어해주면 좋겠어요.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/Logger.java:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * @see <a href="https://datatracker.ietf.org/doc/html/rfc5424">RFC 5242</a>
+ * */
```
```suggestion
 */
```

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/JsonParserTest.java:
```diff
@@ -0,0 +1,30 @@
+
+class JsonParserTest {
+
+    @Test
+    void objToJson_paring_테스트() throws IOException {
+        Task task = new Task(1L, "task1");
+        String parsedTask = JsonParser.toJsonString(Lists.newArrayList(task));
+
+        assertEquals(parsedTask, "{\"id\":1,\"title\":\"task1\"}");
+    }
```
테스트 케이스가 하나인데요, 여러가지 케이스를 더 추가해 봅시다.

Miniminis committed:
refactor 경로 검사로직 PathChecker 클래스로 이동

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/PathCheckerTest.java:
```diff
@@ -0,0 +1,30 @@
+class PathCheckerTest {
+
+    @Test
+    void isMatchedPath_일치할때() {
```
다음번에는 `@DisplayName`을 활용해 보도록 합시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/utils/StringCheckerTest.java:
```diff
@@ -0,0 +1,22 @@
+
+class StringCheckerTest {
+
+    @Test
+    void isEqual_같을때() {
+        boolean result = StringChecker.isEqual("/tasks/1", "/tasks/1");
+
+        Assertions.assertTrue(result);
+    }
+
+    @Test
+    void isEqual_같지_않을때() {
+        boolean result = StringChecker.isEqual("/tasks", "/tasks/1");
+
+        Assertions.assertFalse(result);
+    }
```
테스트 케이스가 좀 더 풍부하면 좋겠어요. null 도 넣어보고, "" 도 넣어보고, 전혀 상관없어 보이는 "_", " " 같은 것들도 넣는 거죠.

johngrib reviewed (approved):
한 주간 고생 많으셨습니다.

Miniminis commented:
> 한 주간 고생 많으셨습니다.

한 주간 리뷰 감사했습니다! 말씀해주신 내용들 2주차 과제에 포함하여 반영해보겠습니다 (꾸벅)
